pkg/vmprovider/providers/vsphere2/vmlifecycle/bootstrap.go

```diff
@@ -319,3 +325,97 @@ func isLinuxGuest(guestID string) bool {
 
 	return strings.Contains(guestID, "Linux") || strings.Contains(guestID, "linux")
 }
+
+func logConfigSpec(
+	vmCtx context.VirtualMachineContextA2,
+	configSpec *vimTypes.VirtualMachineConfigSpec) {
+
+	if !pkgconfig.FromContext(vmCtx).LogFullBootstrapConfigAndCustomizationSpecs {
+		configSpec = SanitizeConfigSpec(configSpec)
+	}
+
+	vmCtx.Logger.Info("Customization Reconfigure", "configSpec", configSpec)
+}
+
+func SanitizeConfigSpec(configSpec *vimTypes.VirtualMachineConfigSpec) *vimTypes.VirtualMachineConfigSpec {
+	cs := *configSpec
+
+	cs.ExtraConfig = slices.Clone(cs.ExtraConfig)
+	for i, ec := range cs.ExtraConfig {
+		optVal := ec.GetOptionValue()
+		if optVal == nil {
+			continue
+		}
+
+		if optVal.Key == constants.CloudInitGuestInfoUserdata {
+			optValCopy := *optVal
+			optValCopy.Value = redacted
+			cs.ExtraConfig[i] = &optValCopy
+			break
+		}
+	}
+
+	if vAppConfig := cs.VAppConfig; vAppConfig != nil && vAppConfig.GetVmConfigSpec() != nil {
+		vmConfigSpec := *vAppConfig.GetVmConfigSpec()
+
+		vmConfigSpec.Property = slices.Clone(vmConfigSpec.Property)
+		for i, vmProp := range vmConfigSpec.Property {
+			if vmProp.Info == nil || vmProp.Info.UserConfigurable == nil || !*vmProp.Info.UserConfigurable {
+				continue
+			}
+
+			info := *vmProp.Info
+			info.Value = redacted
+			vmConfigSpec.Property[i].Info = &info
+		}
+
+		cs.VAppConfig = &vmConfigSpec
+	}
+
+	return &cs
+}
+
+func logCustomizationSpec(
+	vmCtx context.VirtualMachineContextA2,
+	customizationSpec *vimTypes.CustomizationSpec) {
+
+	if !pkgconfig.FromContext(vmCtx).LogFullBootstrapConfigAndCustomizationSpecs {
+		customizationSpec = SanitizeCustomizationSpec(customizationSpec)
+	}
+
+	vmCtx.Logger.Info("Customizing VM", "customizationSpec", customizationSpec)
+}
+
+func SanitizeCustomizationSpec(custSpec *vimTypes.CustomizationSpec) *vimTypes.CustomizationSpec {
+	cs := *custSpec
```

Comment:
This is an exported function and there is an indirect without a nil check. Either make the param by-value or add the nil check.